ff/serving/provider/offline.go

```diff
@@ -106,8 +351,39 @@ func newMemoryOfflineTable() *memoryOfflineTable {
 	}
 }
 
+func (table *memoryOfflineTable) records() []ResourceRecord {
+	allRecs := make([]ResourceRecord, 0)
+	for _, recs := range table.entityMap {
+		allRecs = append(allRecs, recs...)
+	}
+	return allRecs
+}
+
+func (table *memoryOfflineTable) getLastValueBefore(entity string, ts time.Time) interface{} {
+	recs, has := table.entityMap[entity]
+	if !has {
+		return nil
+	}
+	sortedRecs := ResourceRecords(recs)
+	sort.Sort(sortedRecs)
+	lastIdx := len(sortedRecs) - 1
+	for i, rec := range sortedRecs {
+		if rec.TS.After(ts) {
+			// Entity was not yet set at timestamp.
+			if i == 0 {
+				return nil
+			}
+			return sortedRecs[i-1].Value
+		} else if i == lastIdx {
+			// Every record happened before the label.
+			return rec.Value
+		}
+	}
+	panic("Unable to getLastValue before timestamp")
```

Comment:
Do we want this to panic or should it just return an error?

Reply:
or return nil

Reply:
We never get to this line, but I can't return nothing so it is a panic